### PR TITLE
fix: convert past_length to int in _prefill for static cache compatib…

### LINKED ISSUE #46858: _prefill: cache.get_seq_length() returns device tensor, breaks slice arithmetic for static cache + torch.compile on non-CUDA backends
### System Info

  - transformers version: 5.5.3
  - PyTorch version: 2.10.0
  - Python version: 3.10.12
  - OS: Linux (Ubuntu 22.04)
  - Device: CPU (using `torch.compile(backend="inductor")`)
  - Last working version: transformers 5.2.0
  - First broken version: transformers 5.5.3 (where _prefill was introduced)

### Who can help?

@ArthurZucker @gante 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

model = AutoModelForCausalLM.from_pretrained("gpt2")
model.generation_config.cache_implementation = "static"
model.forward = torch.compile(model.forward, backend="inductor")

tokenizer = AutoTokenizer.from_pretrained("gpt2")
input_ids = tokenizer("Hello world", return_tensors="pt").input_ids

# First call: works
output1 = model.generate(input_ids, max_new_tokens=1, do_sample=False)
print("First call OK:", output1.shape)

# Second call: CRASHES
output2 = model.generate(input_ids, max_new_tokens=1, do_sample=False)
print("Second call OK:", output2.shape)

**Note**: On CUDA this may not crash due to CUDA graph replay masking the bug. 
To reproduce on CUDA, you can add a `torch._dynamo.reset()` between calls to force retracing.

### Expected behavior

Multiple `model.generate()` calls with static cache and `torch.compile` should work
without crashing, regardless of the backend device.

## Root Cause Analysis

1. `cache_utils.py` line 383: `self.cumulative_length = torch.tensor([0], dtype=int)`
   - Comment says: "Very important that it's a tensor here, to avoid recompiling"
   
2. `cache_utils.py` line 461: `get_seq_length()` returns this tensor directly
   - Return type annotation says `-> int` but actual return is `torch.Tensor`

3. `generation/utils.py` `_prefill()`: Uses `past_length` in int arithmetic
   - `next_sequence_length = input_ids.shape[1] - past_length` → produces tensor

4. `generation/utils.py` line 502: `prepare_inputs_for_generation(next_sequence_length: int | None)`
   - Type contract violated: receives tensor instead of int

5. `generation/utils.py` line 532: `input_ids[:, -next_sequence_length:]`
   - Tensor-as-slice-index produces wrong (0-length) result

## Suggested Fix

In `_prefill()`, convert `past_length` to Python int after calling `get_seq_length()`:

```python
if (cache := model_kwargs.get("past_key_values")) is not None:
    past_length = cache.get_seq_length()
    if isinstance(past_length, torch.Tensor):
        past_length = int(past_length.item())
```

This is safe because:
- `cumulative_length` remains a tensor on the model (torch.compile compatibility preserved)
- Only the local variable in `_prefill` is converted
- On CUDA (where `get_seq_length()` may already return int): isinstance check is False → no-op
- Restores the `int | None` type contract for `next_sequence_length`

  This worked correctly in transformers 5.2.0 where _prefill() did not exist.
  The regression was introduced when _prefill was added (transformers >= 5.5.0)
  which introduced cache-aware token slicing using cache.get_seq_length().

  In 5.2.0: generate() always passed full input_ids to model.forward (no slicing).
  In 5.5.3: _prefill slices input_ids based on cache.get_seq_length(), but doesn't
  handle the case where get_seq_length() returns a tensor instead of int.

### PR DESCRIPTION
**Pull request by silicon-dispatch-100**
…ility

StaticCache.get_seq_length() returns self.cumulative_length which is intentionally a device tensor (for torch.compile). But _prefill uses it in arithmetic that produces next_sequence_length, which is later passed to prepare_inputs_for_generation(next_sequence_length: int | None).

When next_sequence_length is a tensor instead of int, the slice input_ids[:, -next_sequence_length:] produces wrong (0-length) results, causing model.forward to crash on view(1, 0, -1, 128).

Fix: convert past_length to Python int in _prefill before arithmetic. This preserves cumulative_length as tensor on the model (for compile) while ensuring correct slice behavior in the generation loop.

Fixes #46858

# What does this PR do?

Fixes a bug where `_prefill()` passes a `torch.Tensor` as `next_sequence_length` to
`prepare_inputs_for_generation()` (which expects `int | None`), causing incorrect
slicing that produces 0-length `input_ids` on the second `generate()` call with
static cache.

`StaticLayer.get_seq_length()` returns `self.cumulative_length` — a device tensor by
design (comment: "Very important that it's a tensor here, to avoid recompiling"). But
`_prefill` uses this in int arithmetic and passes the result to 
`prepare_inputs_for_generation(next_sequence_length: int | None)`.

On non-CUDA backends where `torch.compile` retraces on shape changes, the tensor type
causes `input_ids[:, -tensor([4096]):]` to produce `(1, 0)` shaped tensors instead of
`(1, 4096)`.

## Fix

2-line addition in `_prefill()`:
```python
if isinstance(past_length, torch.Tensor):
    past_length = int(past_length.item())
```

## Who is affected?

Any user running `model.generate()` multiple times with:
- `cache_implementation="static"`
- `torch.compile` (any backend except CUDA with `reduce-overhead` which masks the bug)

This includes custom accelerator backends (QAIC, XPU, etc.) and potentially
CPU backend with inductor.

## Tests

- `test_prefill_static_cache_tensor_seq_length`: Verifies no crash on 3 consecutive calls
- `test_prefill_next_sequence_length_type`: Verifies the type contract is maintained

## Related

- `StaticLayer.get_seq_length()` return type annotation says `int` but returns `Tensor`
  (separate issue — the tensor is intentional for torch.compile, but annotation is misleading)

Fixes # (issue)

## Code Agent Policy

The Transformers repo is currently being overwhelmed by a large number of PRs and issue comments written by
code agents. We are currently bottlenecked by our ability to review and respond to them. As a result, 
**we ask that new users do not submit pure code agent PRs** at this time. 
You may use code agents in drafting or to help you diagnose issues. We'd also ask autonomous "OpenClaw"-like agents
not to open any PRs or issues for the moment.

PRs that appear to be fully agent-written will probably be closed without review, and we may block users who do this
repeatedly or maliciously. 

This is a rapidly-evolving situation that's causing significant shockwaves in the open-source community. As a result, 
this policy is likely to be updated regularly in the near future. For more information, please read [`CONTRIBUTING.md`](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md).

- [x] I confirm that this is not a pure code agent PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [x] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by silicon-dispatch-100** [comment]
The CI failures are unrelated to this change. This PR only adds 2 lines
  (isinstance check + int conversion) in `_prefill` which is a no-op when
  `get_seq_length()` already returns an int (as it does on CUDA). The failing
  tests appear to be pre-existing/flaky failures unrelated to generation or
  static cache.

  Happy to rebase if needed once the unrelated failures are resolved.
